internal/dialectadapter/locker.go

```diff
@@ -0,0 +1,128 @@
+
+	"github.com/pressly/goose/v4/internal/dialectadapter/dialectquery"
+	"github.com/sethvargo/go-retry"
+)
+
+var (
+	// defaultLockID is the id used to lock the database for migrations. It is a
+	// crc64 hash of the string "goose". This is used to prevent multiple
+	// goose processes from running migrations at the same time.
+	//
+	// 5887940537704921958
+	defaultLockID = crc64.Checksum([]byte("goose"), crc64.MakeTable(crc64.ECMA))
+
+	// ErrLockNotImplemented is returned when the database does not support locking.
+	ErrLockNotImplemented = errors.New("lock not implemented")
+)
+
+// Locker defines the methods to lock and unlock the database.
+//
+// Locking is an experimental feature and the underlying implementation
+// may change in the future.
+//
+// The only database that currently supports locking is Postgres.
+//
+// Other databases will return ErrLockNotImplemented.
+type Locker interface {
+	// IsLockingEnabled returns true if the database supports locking.
+	IsLockingEnabled() bool
+
+	// LockSession and UnlockSession are used to lock the database for the
+	// duration of a session.
+	//
+	// The session is defined as the duration of a single connection.
+	LockSession(ctx context.Context, conn *sql.Conn) error
+	UnlockSession(ctx context.Context, conn *sql.Conn) error
+
+	// LockTransaction is used to lock the database for the duration of a
+	// transaction.
+	LockTransaction(ctx context.Context, tx *sql.Tx) error
+}
+
+func (s *store) IsLockingEnabled() bool {
+	switch s.querier.(type) {
+	case *dialectquery.Postgres:
+		return true
+	default:
+		return false
+	}
+}
+
+func (s *store) LockSession(ctx context.Context, conn *sql.Conn) error {
+	switch t := s.querier.(type) {
+	case *dialectquery.Postgres:
+		// TODO(mf): need to be VERY careful about the retry logic here to avoid
+		// stacking locks on top of each other. We need to make sure that we
+		// only retry if the lock is not already held.
+		//
+		// This retry is a bit pointless because if we can't get the lock, chances
+		// are another process is holding the lock and we will just spin here
+		// forever. We should probably just remove this retry.
+		//
+		// At best this might help with a transient network issue.
+		return retry.Do(ctx, s.retry, func(ctx context.Context) error {
+			if _, err := conn.ExecContext(ctx, t.AdvisoryLockSession(), defaultLockID); err != nil {
+				return retry.RetryableError(err)
+			}
+			return nil
+		})
+	}
+	return ErrLockNotImplemented
+}
+
+func (s *store) UnlockSession(ctx context.Context, conn *sql.Conn) error {
+	switch t := s.querier.(type) {
+	case *dialectquery.Postgres:
+		return retry.Do(ctx, s.retry, func(ctx context.Context) error {
+			var unlocked bool
+			row := conn.QueryRowContext(ctx, t.AdvisoryUnlockSession(), defaultLockID)
+			if err := row.Scan(&unlocked); err != nil {
+				return retry.RetryableError(err)
+			}
+			if !unlocked {
+				// TODO(mf): provide the user with some documentation on how they can unlock
```

Comment:
This should be added to the documentation as a gotcha, regardless of how edge case this is, when using session-level advisory locks.